pkg/loki/modules.go

```diff
@@ -350,6 +355,72 @@ func (t *Loki) initQueryFrontend() (_ services.Service, err error) {
 	}), nil
 }
 
+func (t *Loki) initRulerStorage() (_ services.Service, err error) {
+	// if the ruler is not configured and we're in single binary then let's just log an error and continue.
+	// unfortunately there is no way to generate a "default" config and compare default against actual
+	// to determine if it's unconfigured.  the following check, however, correctly tests this.
+	// Single binary integration tests will break if this ever drifts
+	if t.cfg.Target == All && t.cfg.Ruler.StoreConfig.IsDefaults() {
+		level.Info(util.Logger).Log("msg", "RulerStorage is not configured in single binary mode and will not be started.")
+		return
+	}
+
+	t.RulerStorage, err = cortex_ruler.NewRuleStorage(t.cfg.Ruler.StoreConfig)
+
+	return
+}
+
+func (t *Loki) initRuler() (_ services.Service, err error) {
+	if t.RulerStorage == nil {
+		level.Info(util.Logger).Log("msg", "RulerStorage is nil.  Not starting the ruler.")
+		return nil, nil
+	}
+
+	t.cfg.Ruler.Ring.ListenPort = t.cfg.Server.GRPCListenPort
+	t.cfg.Ruler.Ring.KVStore.MemberlistKV = t.memberlistKV.GetMemberlistKV
+	q, err := querier.New(t.cfg.Querier, t.cfg.IngesterClient, t.ring, t.store, t.overrides)
+	if err != nil {
+		return nil, err
+	}
+
+	engine := logql.NewEngine(t.cfg.Querier.Engine, q)
+
+	t.ruler, err = ruler.NewRuler(
+		t.cfg.Ruler,
+		engine,
+		prometheus.DefaultRegisterer,
+		util.Logger,
+		t.RulerStorage,
+	)
+
+	if err != nil {
+		return
+	}
+
+	// Expose HTTP endpoints.
+	if t.cfg.Ruler.EnableAPI {
+
+		t.server.HTTP.Handle("/ruler/ring", t.ruler)
+		cortex_ruler.RegisterRulerServer(t.server.GRPC, t.ruler)
+
+		// Ruler Legacy API Routes
+		t.server.HTTP.Path("/api/prom/rules").Methods("GET").Handler(t.httpAuthMiddleware.Wrap(http.HandlerFunc(t.ruler.ListRules)))
```

Comment:
all of our other paths are prefeix /loki/api... I think we should do that for these paths as well

Reply:
good catch